onnxruntime/core/providers/openvino/openvino_execution_provider.cc

```diff
@@ -0,0 +1,571 @@
+  }
+
+  auto opset = model_proto.add_opset_import();
+  opset->set_domain(kOnnxDomain);
+  opset->set_version(graph_viewer.DomainToVersionMap().at(kOnnxDomain));
+  model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
+
+  return model_proto;
+}
+
+//Checks whether the node is supported by OpenVINO
+bool IsOpSupported(std::string name){
+
+    std::set<std::string> supported_ops = {
+        "Add",
+        "BatchNormalization",
+        "Conv",
+        "GlobalAveragePool",
+        "Relu",
+        "Reshape",
+        "Flatten",
+        "Gemm",
+        "MaxPool",
+        "AveragePool",
+        "Concat",
+        "Dropout",
+        "LRN",
+        "Softmax",
+        "Mul",
+        "Sum",
+        "Transpose",
+        "Identity",
+        "MatMul",
+        "Unsqueeze",
+        "ImageScaler",
+        "LeakyRelu",
+        "GlobalMaxPool"};
+
+    auto iter = supported_ops.find(name);
+    return iter != supported_ops.end();
+}
+
+
+//Checks if the entire graph is supported by OpenVINO EP and returns false if it is not.
+
+bool IsGraphSupported(const onnxruntime::GraphViewer& graph_viewer, std::string dev_id){
+
+  auto initializers = graph_viewer.GetAllInitializedTensors();
+
+  auto node_indexes = graph_viewer.GetNodesInTopologicalOrder();
+
+  auto model_proto = GetModelProtoFromFusedNode(graph_viewer);
+
+  auto graph_proto = model_proto.mutable_graph();
+  int input_dims = 0;
+  int output_dims = 0;
+  int num_inputs = graph_viewer.GetInputs().size();
+  int num_outputs = graph_viewer.GetOutputs().size();
+
+  //input_arrays hold the dimensions of graph inputs
+  std::vector<std::vector<int>> input_arrays;
+
+  if (num_inputs != 0)
+    input_dims = graph_proto->input(0).type().tensor_type().shape().dim_size();
+
+  if (num_outputs != 0)
+    output_dims = graph_proto->output(0).type().tensor_type().shape().dim_size();
+
+  //Get input dimensions
+  if (num_inputs != 0) {
+    for (int i = 0; i < num_inputs; i++) {
+      int input_dims_size = graph_proto->input(i).type().tensor_type().shape().dim_size();
+      std::vector<int> temp_arr;
+
+      for (int j = 0; j < input_dims_size; j++) {
+        temp_arr.push_back(graph_proto->input(i).type().tensor_type().shape().dim(j).dim_value());
+      }
+      input_arrays.push_back(temp_arr);
+    }
+  }
+
+  //GPU Plugin does not support single dimensional input and 5 dimensional input
+  if (dev_id == "GPU") {
+    if (input_dims == 1 || input_dims == 5 || output_dims == 5)
+      return false;
+  }
+
+  bool op_sum = false;
+  bool op_gemm = false;
+
+  for (auto index : node_indexes) {
+    auto node = graph_viewer.GetNode(index);
+
+
+    //Check if the Operation is Supported by OpenVINO
+    if (!IsOpSupported(node->OpType())) {
+      return false;
+    }
+
+    if (node->OpType() == "Sum") {
+      op_sum = true;
+    }
+
+    if (node->OpType() == "Gemm") {
+      op_gemm = true;
+    }
+
+    //BatchNorm, Conv and Reshape cant take more than 1 input
```

Comment:
would be nice if we could keep logic for a single op together. there's some ops that have logic in multiple places.   
see ngraph as example:
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/ngraph/ngraph_execution_provider.cc#L102

Also consider using helper functions for repeated code blocks.

Reply:
Hey George, 
As there are a lot of operators with the same condition i thought that it would be better to group them by condition. Would it be better to group them by op?

Agreed. Will use a helper function.

Reply:
isn't it easier to understand the restrictions if it's grouped by op?

Reply:
Yes maybe for understanding the code it might be a little confusing to keep track of why an operator is not supported if it has conditions in multiple places.